<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [100]:
import pandas as pd
import psycopg2


# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [103]:
# Для определения количества вакансий в БД, произведем подсчет по столбцу ID из таблицы Vacancies.
query_3_1 = f'''select 
               count(id) 
            from vacancies 
'''

In [104]:
# Количество вакансий в БД.
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/3025775339.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [105]:
# Для определения количества работодателей в БД, произведем подсчет по столбцу ID из таблицы Employers.
query_3_2 = f'''select 
               count(id) 
            from employers 
'''

In [106]:
# Количество работодателей
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/1747573193.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [107]:
# Для определения количества регионов, где размещены вакансии, произведем подсчет по столбцу ID из таблицы Areas.
query_3_3 = f'''select 
               count(id) 
            from areas 
'''

In [108]:
# Количество регионов
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/513417693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [109]:
#Для определения количества сфер деятельности работодателей в БД, произведем подсчет по столбцу ID из таблицы Industries.
query_3_4 = f'''select 
               count(id) 
            from industries 
'''

In [110]:
# Количество сфер деятельности работодателей.
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/3318631973.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


count
0    294

***

#ВЫВОДЫ ПО ПРЕДВАРИТЕЛЬНОМУ АНАЛИЗУ ДАННЫХ

Предварительный анализ вакансий показал следующее:

    1. Количество вакансий, размещенных на сайте HeadHunter, равно 49197 шт.
    2. Разместили свои объявления 23501 работодатель.
    3. География распределения вакансий насчитывает 1362 региона, это не только Россия, но и другие страны.
    4. В 294 сферах деятельности работодателей, ведется поиск соискателей.


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [111]:
#Для подсчета количества вакансий в каждом регионе, необходимо произвести соединение таблиц Vacancies и Areas. Соединение проведем 
#по ID региона вакансии из таблицы Vacancies и по ID идентификатору города из таблицы Areas. Сгруппируем по названию города и 
# отсортируем по убыванию количества вакансий.
query_4_1 = f'''select 
                  a.name ,
                  count(v.id) cnt
                from vacancies v
                  join areas a on v.area_id=a.id
                group by a.name
                order by cnt desc
'''

In [112]:
# Количество вакансий в каждом регионе
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/1084359835.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


name   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [113]:
#Подсчета количества вакансий, у которых заполнено хотя бы одно из двух полей с Зарплатой, произведем по столбцу ID из таблицы
# Vacancies. Также пропишем условия, где столбец salary_from или salary_to не должны равняться 0.
query_4_2 = f'''select 
                  count(id) cnt
                from vacancies 
                where salary_from is not null or salary_to is not null
'''

In [114]:
# Количество вакансий, у которых заполнено хотя бы одно поле с зарплатой.
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/2133769361.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [115]:
#Для данного запроса рассчитаем среднее в столбцах salary_from, salary_to из таблицы Vacancies и округлим результат до целого числа.
query_4_3 = f'''select 
                  round(avg(salary_from)) average_lower_bound,
                  round(avg(salary_to)) average_upper_bound
                from vacancies 
'''

In [116]:
# Среднее занчение для нижней и верхней границы зарплатной вилки
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/216816730.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


average_lower_bound  average_upper_bound
0              71065.0             110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [117]:
#Для получения уникальных сочетаний из столбцов типа рабочего графика Schedule и типа трудоустройства Employment, используем Distinct. 
# Подсчет проводим по столбцу ID из таблицы Vacancies. Сгруппируем по столбам schedule и employment. Сортировка по убыванию по 
# количеству вакансий.
query_4_4 = f'''select distinct
                  count(id) cnt,
                  v.schedule schedule,
                  v.employment employment
                from vacancies v
                group by 2,3
                order by cnt desc 
'''

In [118]:
# Количество вакансий для каждого сочетания типа рабочего графика  и типа трудоустройства 
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/2523917370.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


cnt          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [119]:
#Для запроса используем столбцы ID и Experience из таблицы Vacancies. Считаем по ID. Группировка по опыту работы. 
# Сортировка по возрастанию количества вакансий.
query_4_5 = f'''select 
                  experience,
                  count(id) cnt
                from vacancies 
                group by 1
                order by cnt 
'''

In [120]:
# Количество вакансий, в которых ребуется Опыт работы
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/3084393552.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

ВЫВОДЫ ПО ДЕТЕЛЬНОМУ АНАЛИЗУ ДАННЫХ

1. Рассмотрим результаты распределения количества вакансий по регионам, возьмем ТОП-5 регионов:

   1.1. В Москве самое большое количество размещенных объявлений - 5333 шт. (это 10,84% от всех размещенных вакасний).

   1.2. В Санкт-Петербурге разместили  - 2851 шт. (это 5,8% ).

   1.3. На 3-ем месте находится Минск  - 2112 шт. (это 4,3% ).

   1.4. В Новосибирске разместили - 2006 шт. (это 4,1% ).

   1.5. На 5-ом месте находится Алматы - 1892 шт. (это 3,85% ).
   
Из цифр видно, что 2 города из ТОП-5 являются столицами  России и Белоруси. 3 других города - являются городами-миллионниками России и Казахстана, население данных городов прилижается к 2 миллионам. Логично, что здесь размещено наибольшее количество вакансий.

2. Проанализируем вилку зарплат, предлагаемых в вакансиях.

   2.1. Среднее значение минимальной предлагаемой зарлаты составляет 71065 рублей, а среднее максимального значения - 110537 рублей.
Можно предположить, что среднее значение максимально предлагаемой зарплаты, распределено в крупных регионах. И наоборот, в маленьких регионах предлагают меньшую зарплау.

3. Выведем ТОП-5 популярныч сочетаний Графика работы и Занятости. 

   3.1. Полная занятость и Полный рабочий день - 35367 вакансий (это 71,9 % от всех размещенных вакасний).

   3.2. Полная занятость и Удаленная работа - 7802 вакансий (это 15,9 %).

   3.3. Полная занятость и Гибкий график - 1593 вакансий (это 3,2 % ).

   3.4. Частичная занятость и Удаленная работа - 1312 вакансий (это 2,7 %).

   3.5. Полная занятость и Сменный график - 940 вакансий (это 1,9 %).

Видим, что основная масса вакансий около 93% - предлагают полную занятость и различные варианты графика работы. Остальные сочетания имеют минимальные значения от общего количества вакансий.

4. Обратимся к критерию Опыт работы:

   4.1. Нет Опыта работы - 7197 шт. (это 14,6 % от всех размещенных вакасний). 

   4.2. Опыт работы от 1 до 3 лет - 26153 шт. (это 53,2 %). 

   4.3. Опыт работы от 3 до 6 лет - 14511 шт. (это 29,5 %). 

   4.4. Опыт работы от 6 и более лет - 1337 шт. (это 2,7 %). 
   
Наиболее востребованный Опыт работы - от 1 до 3 лет - более 53% вакансий, далее идет снижение количества вакансий. 2,7% вакансий относится к Опыту работы  от 6 лет и более лет.

***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [121]:
# Для данного запроса необходимо произвести соединение таблиц Vacancies и Employers. Соединение проведем 
#по ID номеру работодателя из таблицы Employers и по Employer_id идентификатору работодателя из таблицы Vacancies. 
# Столбцы к выводу employers_name и cnt. Сгруппируем по Названию работодателя и отсортируем по убыванию количества вакансий.
# Выведем Топ-5 работодателей по количеству вакансий.
query_5_1 = f'''select 
                  e.name employers_name,
                  count(v.id) cnt
                from vacancies v
                  join employers e on e.id=v.employer_id
                group by e.name
                order by cnt desc
                limit 5
'''


In [122]:
# Топ-5 Работодателей по количеству вакансий
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/2865110728.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


employers_name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [123]:
# Для данного запроса необходимо произвести соединение таблиц Areas, Vacancies и Employers через left join, т.к. перед добавлением 
# данных из Правых таблиц, проводится проверка на соответствие условий из Левой таблицы. Соединение проведем по ID идентификатора 
# города из таблицы Areas, по Area регион регистрации из таблицы Employers и по Area_id идентификатору региона вакансии из таблицы 
# Vacancies. Столбцы к выводу areas_name,cnt_employers и cnt_vacancies. Добавим условие, что подсчет будет производиться по регионам,
# где нет вакансий. Сгруппируем по Названию региона и отсортируем по убыванию количества работодателей.
query_5_2 = f'''select 
                  a.name areas_name,
                  count(e.id) cnt_employers,
                  count(v.id) cnt_vacancies
                from areas a
                  left join employers e on a.id=e.area
                  left join vacancies v on v.area_id=a.id
                where v.id is null
                group by a.name
                order by cnt_employers desc               
'''

In [124]:
# Количество вакансий и работодателей для каждого региона
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/3829272566.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


areas_name  cnt_employers  cnt_vacancies
0                            Россия            410              0
1                         Казахстан            207              0
2                Московская область             75              0
3                Краснодарский край             19              0
4                          Беларусь             18              0
..                              ...            ...            ...
588                        Кувандык              0              0
589                        Куйбышев              0              0
590  Куйбышево (Ростовская область)              0              0
591                        Култаево              0              0
592                        Кульсары              0              0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [125]:
# Для данного запроса необходимо произвести соединение таблиц Vacancies и Employers через left join. Соединение проведем по ID 
# номера работодателя из таблицы Employers и по Employer_id идентификатору аботодателя из таблицы Vacancies. Столбцы к выводу 
# employers_name и cnt_areas.Сгруппируем по Идентификатору работодателя и отсортируем по убыванию количества регионов.
query_5_3 = f'''select 
                  e.name employers_name,
                  count(distinct v.area_id) cnt_areas
                from employers e
                  left join vacancies v on e.id=v.employer_id
                group by e.id
                order by cnt_areas desc
'''               

In [126]:
# Количество регионов, в которых есть вакансии у каждого работодателя
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/3226416858.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


employers_name  cnt_areas
0                      Яндекс        181
1                  Ростелеком        152
2                  Спецремонт        116
3      Поляков Денис Иванович         88
4                   ООО ЕФИН          71
...                       ...        ...
23496              Но Смокинг          0
23497        МАСТЕРСКАЯ CREED          0
23498           Dzing Finance          0
23499                    КОРС          0
23500             Наука-Связь          0

[23501 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [127]:
# Для данного запроса необходимо произвести соединение таблиц Vacancies и Employers_industries через left join. Соединение проведем по 
# ID номера работодателя из таблицы Employers и по Employer_id идентификатору аботодателя из таблицы Employers_industries. Добавляем 
# условия, что столбец Indusry_id (идентификатор сферы деятельности) должен равняться 0.
query_5_4 = f'''select 
                  count(e.id) cnt
                from employers e
                   left join employers_industries ei on e.id=ei.employer_id
                where ei.industry_id is null
'''

In [128]:
# Количество работодателей, у которых не указана сфера деятельности
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/2218440283.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


cnt
0  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [129]:
# Для данного запроса необходимо произвести соединение таблиц Employers и Employers_industries через left join. Соединение проведем по 
# ID номера работодателя из таблицы Employers и по Employer_id идентификатору аботодателя из таблицы Employers_industries. Добавляем 
# условие, что количество из столбца Indusry_id (идентификатор сферы деятельности) должен равняться 4. Группируем по Названию 
# работодателя и сортируем в Алфавитном порядке по Названию работодателя. Берем первые три компании.
query_5_5 = f'''select 
                  e.name employers
                from employers e
                   left join employers_industries ei on e.id=ei.employer_id
                group by employers
                having count(ei.industry_id) = 4
                order by employers
                limit 3
'''


In [130]:
# Топ-3 компаний, у которых указано 4 сферы деятельности
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/1466586871.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


employers
0  101 Интернет
1      21vek.by
2          2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [132]:
# Для данного запроса необходимо произвести соединение таблиц Industries и Employers_industries. Соединение проведем по 
# ID сферы деятельности из таблицы Industries и по Employer_id идентификатору аботодателя из таблицы Employers_industries. Добавляем
# условие о том, что название сферы деятельности компании для подсчета - 'Разработка программного обеспечения'.
query_5_6 = f'''select 
                  count(ei.employer_id) cnt
                from employers_industries ei
                  join industries i on i.id=ei.industry_id
                where i.name = 'Разработка программного обеспечения'
'''

In [133]:
# Количество работодателей, у которых указана сфера деятельности Разработка ПО. 
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/1267939077.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


cnt
0  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [134]:
#Для чтения HTML-таблицы с городами-миллионниками, из статьи в Википедии, используем функцию библиотеки Pandas - read_html(). Указываем  
#номер нужной таблицы на странице. И выбираем нужный нам столбец, преобразовывая его в список.
df= pd.read_html(
    'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')[0]

million_cities=df.iloc[:,1]


In [135]:
# Для данного запроса необходимо произвести соединение таблиц Areas, Vacancies и Employers через left join. Соединение проведем по ID 
# идентификатора города из таблицы Areas, по Area_id регион вакансии из таблицы Vacancies, по Employer_id идентификатору работодателя
# из таблицы Vacancie и по Id работодателя из таблицы Employers. Столбцы к выводу city_million и cnt. Добавим условия, что нас 
# интересует компания Яндекс и регионы размещения вакансий Города-миллионники.Сгруппируем по Названию региона и отсортируем по убыванию
# количества вакансий. Через объединение union all добавим строку Total с итоговым количеством вакансий компании.
query_5_7 = f'''(select 
                  a.name city_million,
                  count(v.id) cnt
                from employers e
                  left join vacancies v on e.id=v.employer_id
                  left join areas a on v.area_id=a.id
                where e.name = 'Яндекс' and a.name  in {million_cities}
                group by 1
                order by 2
                union all
                select
                   'Total', count(v.id)
                from vacancies v
'''

In [136]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/2595257991.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


DatabaseError: Execution failed on sql '(select 
                  a.name city_million,
                  count(v.id) cnt
                from employers e
                  left join vacancies v on e.id=v.employer_id
                  left join areas a on v.area_id=a.id
                where e.name = 'Яндекс' and a.name  in 0              Москва
1     Санкт-Петербург
2         Новосибирск
3        Екатеринбург
4              Казань
5          Красноярск
6     Нижний Новгород
7           Челябинск
8                 Уфа
9              Самара
10     Ростов-на-Дону
11          Краснодар
12               Омск
13            Воронеж
14              Пермь
15          Волгоград
Name: Город, dtype: object
                group by 1
                order by 2
                union all
                select
                   'Total', count(v.id)
                from vacancies v
': syntax error at or near "0"
LINE 7: ...           where e.name = 'Яндекс' and a.name  in 0         ...
                                                             ^


***

ВЫВОДЫ ПО АНАЛИЗУ РАБОТОДАТЕЛЕЙ

1. Топ-5 работодателей по количеству размещенных вакансий:

   1.1. Яндекс - 1933 (3.93% от всех размещенных вакансий)

   1.2. Ростелеком - 491 (1%)

   1.3. Тинькофф - 444 (0.9%)

   1.4. СБЕР - 428 (0.87%)

   1.5. Газпром нефть - 331 (0.67%)

Комания Яндекс имеет много сфер деятельности от IT-направления до доставки еды, такси и обучения. Данными услугами пользуются во многих регионах, поэтому такое количество вакансий. По сравнению со второй строкой (Ростелеком) рейтинга разница по количеству почти в 4 раза. Остальные из ТОП-5 имеют скромные цифры. 

2. ТОП-3 регионов, в которых нет вакансий, но  в которых находится  наибольшее количество работодателей

   2.1. Россия - 410 (1.75% от общего числа работодателей)

   2.2. Казахстан - 207 (0.9%)

   2.3. Московская область - 75 (0.32%)

3. ТОП-5 работодателей, у которых самое большое количество регионов, где они разместили вакансии

   3.1. Яндекс - 181 (13.3% от общего количества регионов)

   3.2. Ростелеком - 152 (11.2%)

   3.3. Спецремонт - 116 (8.52%)

   3.4. Поляков Денис Иванович - 88 (6.5%)

   3.5. ООО ЕФИН - 71 (5.21%)

Яндекс здесь также занимает первое место из-за географии оказываемых услуг и их количества. Также востребованы услуги домашнего ТВ и интернета. Третье место занимает компания, занимающаяся ремонтом бытовой техники и представлена в 116 регионах. На 4 месте компания, которая занимается работой с данными(сбор,обработка,создание БД). И на 5 месте компания связанная, с разработкой и внедрением технологий в банковском и финансовом секторах.

4. У 8419 (35.9%) работодателей не указана  сфера деятельности, что может сказаться на выводах.

5. Рассмотрим  ТОП-3 компании, где количество указанных сфер деятельности  - 4.

   5.1. 101.интерет - 410 (1.75% от общего числа работодателей)

   5.2. Казахстан - 207 (0.9%)
   
   5.3. Московская область - 75 (0.32%)

6. Количество компаний, занимающихся Разработкой ПО - 3553 - это 15,2% от общего количества работодателей.

7. 


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# Для подсчета необходимоо применить count к таблице Vacancies. Добавляем условие о том, что в названии вакансии есть слова 'данн' или
#'data', предварительно приведя название вакансии к нижнему регистру.
query_6_1 = f'''select 
                  count(*)
                from vacancies 
                where lower(name) like '%data%' or lower(name) like '%данн%'
'''

In [ ]:
# Количество вакансий, которые имеют отношения к данным
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_4827/1281472032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# Для подсчета необходимоо применить count к таблице Vacancies. Добавляем условия о том, что в описании вакансии есть словосочетания,
# которые можно отнести к специальности дата-сайентистов. При помощи метода ILIKE, который не чувствителен к регистру, сравним строки и 
#добавим 3 условия для Junior специалиста.
query_6_2 = f'''select 
                  count(*)
                from vacancies 
                where (name ILIKE '%data scientist%' 
                  or name ILIKE '%data science%' 
                  or name ILIKE '%исследователь данных%' 
                  or (name ILIKE '%ML%' and name not ILIKE '%HTML%') 
                  or name ILIKE '%machine learning%' 
                  or name ILIKE '%машинн%обучен%') 
                  and (name ILIKE 'junior' 
                  or experience ILIKE 'Нет опыта' 
                  or employment ILIKE 'Стажировка')
'''

In [ ]:
# Количество вакансий уровня Junior
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_4827/118162426.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


count
0     28

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# Для подсчета необходимоо применить count к таблице Vacancies. Добавляем условия о том, что в описании вакансии есть словосочетания,
# которые можно отнести к специальности дата-сайентистов. При помощи метода ILIKE, который не чувствителен к регистру, сравним строки. 
#Добавим  условие, что в перечне навыков есть 'SQL' или 'postgres'.
query_6_3 = f'''select 
                  count(*)
                from vacancies 
                where ((name ILIKE '%data scientist%' 
                  or name ILIKE '%data science%' 
                  or name ILIKE '%исследователь данных%' 
                  or (name ILIKE '%ML%' and name not ILIKE '%HTML%') 
                  or name ILIKE '%machine learning%' 
                  or name ILIKE '%машинн%обучен%') 
                  and (key_skills ILIKE '%SQL%' or key_skills ILIKE '%postgres%'))
'''

In [ ]:
# Количество вакансий, с навыками SQL и postgres
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3


/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_4827/3535121211.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# Для подсчета необходимоо применить count к таблице Vacancies. Добавляем условия о том, что в описании вакансии есть словосочетания,
# которые можно отнести к специальности дата-сайентистов. При помощи метода ILIKE, который не чувствителен к регистру, сравним строки.  
#Добавим  условие, что в перечне навыков есть 'Python'.
query_6_4 = f'''select 
                  count(*)
                from vacancies 
                where ((name ILIKE '%data scientist%' 
                  or name ILIKE '%data science%' 
                  or name ILIKE '%исследователь данных%' 
                  or (name ILIKE '%ML%' and name not ILIKE '%HTML%') 
                  or name ILIKE '%machine learning%' 
                  or name ILIKE '%машинн%обучен%') 
                  and (key_skills ILIKE '%Python%'))
'''

In [ ]:
# Количество вакансий, с ключевым навыком Python
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_4827/3953005136.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# Для подсчета необходимо в key_skills из таблицы Vacancies заменить символ 'CHR9' на пробел, при помощи метода Replace. Произведем 
# вычитание длины столбца key_skills до преобразования и после преобразования. Возьмем среднее значение разницы через AVG и округлим
#получившийся результат.Добавляем условия о том, что в описании вакансии есть словосочетания,которые можно отнести к специальности 
# дата-сайентистов. При помощи метода ILIKE, который не чувствителен к регистру, сравним строки.  Добавим  условие, что в перечне 
# навыков нет пропуском.
query_6_5 = f'''select 
                  round(avg(len(key_skills))-len(replace(key_skills,'CHR9', '')),2)
                from vacancies v
                where (name ILIKE '%data scientist%' 
                  or name ILIKE '%data science%' 
                  or name ILIKE '%исследователь данных%' 
                  or (name ILIKE '%ML%' and name not ILIKE '%HTML%') 
                  or name ILIKE '%machine learning%' 
                  or name ILIKE '%машинн%обучен%' 
                  and key_skills is not null )
'''

In [ ]:
# Количество вакансий, с ключевым навыком Python
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_4827/2426702491.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


DatabaseError: Execution failed on sql 'select 
                  round(avg(len(key_skills))-len(replace(key_skills,'CHR9', '')),2)
                from vacancies v
                where (name ILIKE '%data scientist%' 
                  or name ILIKE '%data science%' 
                  or name ILIKE '%исследователь данных%' 
                  or (name ILIKE '%ML%' and name not ILIKE '%HTML%') 
                  or name ILIKE '%machine learning%' 
                  or name ILIKE '%машинн%обучен%' 
                  and key_skills is not null )
': function len(text) does not exist
LINE 2:                   round(avg(len(key_skills))-len(replace(key...
                                    ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
#Так как значения в зарплате могут быть не заполнены, но они нужны для расчета средней зарплаты, необходимо использовать функцию 
# coalesce. Она возвращает первый попавшийся аргумент, отличный от NULL. Произведем вычитание длины столбца key_skills до преобразования и длину после преобразования.Для преобразования необходимо в key_skills из 
# таблицы Vacancies заменить символ 'CHR9' на пробел, при помощи метода Replace. Возьмем среднее значение разницы через AVG и
# округлим получившийся результат. Добавляем условия о том, что в описании вакансии есть словосочетания,которые можно отнести к 
# специальности дата-сайентистов. При помощи метода ILIKE, который не чувствителен к регистру, сравним строки.  Добавим  условие, что 
# в перечне навыков нет пропуском. Сгруппируем по опыту работы и отсортируем по убыванию средней зарплаты.
query_6_6 = f'''select 
                  experience,
                  round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)),0) salary_average
                from vacancies v
                where (name ILIKE '%data scientist%' 
                  or name ILIKE '%data science%' 
                  or name ILIKE '%исследователь данных%' 
                  or (name ILIKE '%ML%' and name not ILIKE '%HTML%') 
                  or name ILIKE '%machine learning%' 
                  or name ILIKE '%машинн%обучен%'
                  and (salary_from is not null or salary_to is not null))
                  group by experience
                  order by 2 desc
'''

In [ ]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_4827/2297777544.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


experience  salary_average
0       От 3 до 6 лет        256454.0
1         Более 6 лет        157933.0
2  От 1 года до 3 лет        144233.0
3           Нет опыта         74643.0

ВЫВОДЫ ПО ПРЕДМЕТНОМУ АНАЛИЗУ

1. Количество вакансий, в которых упоминается работа с Данными - 1771 или 3,6% от общего количества вакансий. 

2. Вакансии в сфере DS уровня Junior - , это % (от количества вакансий по работе с Данными).

3. 229 вакансий (13%), в которых указаны сразу 2 навыка: SQL и Postgres.

4. С ключевым навыком Python набралось - 357 вакансий (20.1%).

5. Среднее количество навыков, предъявляемых для DS-специалистов - шт.

6. Рассмотрим среднюю зарплату для DS-специалистов, в зависимости от требуемого опыта работы.

***

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения 
# исследования.


In [ ]:
# Закрывам  соединение после окончания работы
connectio.close()

In [138]:
# Для данного запроса необходимо произвести соединение таблиц Employers и Employers_industries через join. Соединение проведем по 
# ID номера работодателя из таблицы Employers и по Employer_id идентификатору работодателя из таблицы Employers_industries.  Группируем 
# по Названию работодателя и сортируем по убыванию количества сфер деятельности. Берем первые 10 компании.
query_7_1 = f'''select 
                  e.name employers,
                  count(ei.industry_id)
                from employers e
                  join employers_industries ei on e.id=ei.employer_id 
                group by employers
                order by 2 desc
                limit 10
'''
df_7_1 = pd.read_sql_query(query_7_1, connection)
df_7_1

/var/folders/qq/6msm4p1n01q2qy3j8mhcrmz40000gn/T/ipykernel_82956/2701527773.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_1 = pd.read_sql_query(query_7_1, connection)


employers  count
0            Модуль     16
1   Простые решения     11
2           Энергия     10
3             Прайд     10
4               СБК      9
5             Альфа      9
6           Эксперт      9
7  Бизнес-Ассистанс      8
8            Уфанет      8
9        Автомастер      8